extensions-contrib/lookup-aware-server-selector/src/main/java/org/apache/druid/client/selector/filter/lookup/LookupAwareFilterStrategy.java

```diff
@@ -0,0 +1,181 @@
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+@JsonTypeName("lookupsLoaded")
+public class LookupAwareFilterStrategy implements ServerFilterStrategy
+{
+  public static final String LOOKUPS_CONTEXT_KEY = "lookups";
+  public static final String QUERY_LOOKUPS_CACHE_KEY = "__lookups";
+  private final LookupStatusView lookupStatusView;
+
+  @JsonCreator
+  public LookupAwareFilterStrategy(@JacksonInject LookupStatusView lookupStatusView)
+  {
+    this.lookupStatusView = lookupStatusView;
+  }
+
+  @Override
+  public <T> Set<QueryableDruidServer> filter(Query<T> query, Set<QueryableDruidServer> servers)
+  {
+    if (query == null) {
+      return servers;
+    }
+    final Set<String> lookups;
+    if (query.getContext() != null
+        && query.getContext().containsKey(QUERY_LOOKUPS_CACHE_KEY)) {
+      lookups = (Set<String>) query.getContext().get(QUERY_LOOKUPS_CACHE_KEY);
+    } else {
+      lookups = new HashSet<String>();
+      findLookupsInQuery(query, lookups);
+      if (query.getContext() != null) {
+        try {
+          query.getContext().put(QUERY_LOOKUPS_CACHE_KEY, lookups);
+        }
+        catch (UnsupportedOperationException e) {
+          // TODO This is for immutable contexts, what to do with it?
+        }
+      }
+    }
+    if (lookups.size() == 0) {
+      return servers;
+    }
+
+    Set<QueryableDruidServer> filteredServers = servers.stream()
+        .filter(server -> lookupStatusView.hasAllLookups(server.getServer().getHostAndPort().toString(), lookups))
+        .collect(Collectors.toSet());
+
+    if (filteredServers.size() == 0) {
+      throw new ISE("Unable to find servers with all lookups [%s]", lookups.stream().collect(Collectors.joining(",")));
+    }
+
+    return filteredServers;
+  }
+
+  private <T> Set<String> findLookupsInQuery(Query<T> query, Set<String> lookupsInQuery)
+  {
+    if (query.getContext() != null && query.getContext().containsKey(LOOKUPS_CONTEXT_KEY)) {
+      Object userLookups = query.getContext().get(LOOKUPS_CONTEXT_KEY);
+      if (userLookups instanceof Collection) {
+        lookupsInQuery.addAll(((Collection) userLookups));
+      }
+    }
+
+    findLookupsInDataSources(query.getDataSource(), lookupsInQuery);
+    findLookupsInDimensions(query, lookupsInQuery);
+    findLookupsInVirtualColumns(query, lookupsInQuery);
+
+    return lookupsInQuery;
+  }
+
+  public <T> void findLookupsInVirtualColumns(Query<T> query, Set<String> lookupsInQuery)
+  {
+    VirtualColumns virts = query.getVirtualColumns();
+    for (VirtualColumn vc : virts.getVirtualColumns()) {
+      if (vc instanceof ExpressionVirtualColumn) {
+        ExpressionVirtualColumn evc = (ExpressionVirtualColumn) vc;
+        Expr expr = evc.getParsedExpression().get();
+        Matcher m = Pattern.compile("lookup\\(.*?, \\'(.*?)\\'\\)").matcher(expr.stringify());
```

Comment:
I don't like using a regex here. Ideally, LookupExprMacro.LookupExpr would be promoted from a local class to something accessible - but I didn't want to include any other changes to core Druid on this PR, if the other commit gets into core then the rest of this functionality can be achieved just with this extension.